File: apps/web-app/src/types/subscription.ts
import type { Delegation } from '@metamask/delegation-toolkit';
import type { ProductResponse, ProductTokenResponse } from './product';

/**
 * Subscription status types
 */
export type SubscriptionStatus = 'active' | 'canceled' | 'past_due' | 'expired';

/**
 * Request payload for creating a subscription (matches Go backend SubscribeRequest)
 */
export interface SubscribeRequest {
  subscriber_address: string;
  product_id: string;
  product_token_id: string;
  token_amount: string;
  delegation: Delegation;
  addons?: SubscriptionAddonRequest[];
}

/**
 * Subscription addon request
 */
export interface SubscriptionAddonRequest {
  product_id: string;
  quantity: number;
}

/**
 * Customer information within a subscription
 */
export interface SubscriptionCustomer {
  id: string;
  num_id: number;
  name?: string;
  email: string;
  phone?: string;
  description?: string;
  finished_onboarding: boolean;
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

/**
 * Subscription line item response
 */
export interface SubscriptionLineItemResponse {
  id: string;
  subscription_id: string;
  product_id: string;
  line_item_type: 'base' | 'addon';
  quantity: number;
  unit_amount_in_pennies: number;
  currency: string;
  price_type: string;
  interval_type?: string;
  total_amount_in_pennies: number;
  is_active: boolean;
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  product?: ProductResponse;
}

/**
 * Represents a subscription along with its associated price and product details (new, matches Go backend SubscriptionResponse)
 */
export interface SubscriptionResponse {
  id: string;
  num_id: number;
  workspace_id: string;
  customer_id?: string;
  customer_name?: string;
  customer_email?: string;
  customer?: SubscriptionCustomer;
  status: SubscriptionStatus;
  current_period_start: string;
  current_period_end: string;
  next_redemption_date?: string;
  total_redemptions: number;
  total_amount_in_cents?: number;
  token_amount: number;
  delegation_id: string;
  customer_wallet_id?: string;
  external_id?: string;
  payment_sync_status?: string;
  payment_synced_at?: string;
  payment_sync_version?: number;
  payment_provider?: string;
  initial_transaction_hash?: string;
  metadata?: {
    wallet_address?: string;
    [key: string]: unknown;
  };
  created_at: string;
  updated_at: string;
  trial_start?: string;
  trial_end?: string;
  canceled_at?: string;
  cancel_at?: string; // Scheduled cancellation date
  cancellation_reason?: string;
  paused_at?: string;
  pause_ends_at?: string;
  product: ProductResponse;
  product_token: ProductTokenResponse;
  line_items?: SubscriptionLineItemResponse[];
  scheduled_changes?: ScheduledChange[];
}

/**
 * Line item update for subscription changes
 */
export interface LineItemUpdate {
  action: 'add' | 'update' | 'remove';
  line_item_id?: string;
  product_id?: string;
  price_id?: string;
  product_token_id?: string;
  quantity: number;
  unit_amount?: number;
}

/**
 * Request types for subscription management
 */
export interface UpgradeSubscriptionRequest {
  line_items: LineItemUpdate[];
  reason?: string;
}

export interface DowngradeSubscriptionRequest {
  line_items: LineItemUpdate[];
  reason?: string;
}

export interface CancelSubscriptionRequest {
  reason: string;
  feedback?: string;
}

export interface PauseSubscriptionRequest {
  pause_until?: string; // ISO 8601 timestamp
  reason?: string;
}

export interface PreviewChangeRequest {
  change_type: 'upgrade' | 'downgrade';
  line_items: LineItemUpdate[];
}

/**
 * Proration calculation result
 */
export interface ProrationResult {
  credit_amount: number;
  charge_amount: number;
  net_amount: number;
  days_total: number;
  days_used: number;
  days_remaining: number;
  old_daily_rate: number;
  new_daily_rate: number;
  calculation: Record<string, any>;
}

/**
 * Change preview response
 */
export interface ChangePreview {
  current_amount: number;
  new_amount: number;
  proration_credit?: number;
  immediate_charge?: number;
  effective_date: string;
  proration_details?: ProrationResult;
  message?: string;
}

/**
 * Scheduled change information
 */
export interface ScheduledChange {
  id: string;
  subscription_id: string;
  change_type: 'upgrade' | 'downgrade' | 'cancel' | 'pause' | 'resume';
  scheduled_for: string;
  from_line_items?: any;
  to_line_items?: any;
  proration_amount_cents?: number;
  proration_calculation?: any;
  status: 'scheduled' | 'processing' | 'completed' | 'cancelled' | 'failed';
  reason?: string;
  initiated_by?: string;
  processed_at?: string;
  created_at: string;
  updated_at: string;
}

/**
 * Subscription state history entry
 */
export interface SubscriptionStateHistory {
  id: string;
  subscription_id: string;
  from_status?: SubscriptionStatus;
  to_status: SubscriptionStatus;
  from_amount_cents?: number;
  to_amount_cents?: number;
  line_items_snapshot?: any;
  change_reason?: string;
  schedule_change_id?: string;
  occurred_at: string;
}

File: libs/go/types/api/responses/subscription.go
package responses

import (
	"time"

	"github.com/cyphera/cyphera-api/libs/go/db"
	"github.com/cyphera/cyphera-api/libs/go/types/api/params"
	"github.com/google/uuid"
)

// ProcessSubscriptionResult represents the result of processing a subscription
type ProcessSubscriptionResult struct {
	TransactionHash string    `json:"transaction_hash"`
	PaymentID       uuid.UUID `json:"payment_id"`
	Success         bool      `json:"success"`
	Error           string    `json:"error,omitempty"`
}

// ProcessDueSubscriptionsResult represents the result of processing all due subscriptions
type ProcessDueSubscriptionsResult struct {
	ProcessedCount   int         `json:"processed_count"`
	SuccessfulCount  int         `json:"successful_count"`
	FailedCount      int         `json:"failed_count"`
	ProcessedIDs     []uuid.UUID `json:"processed_ids"`
	FailedIDs        []uuid.UUID `json:"failed_ids"`
	ProcessingErrors []string    `json:"processing_errors,omitempty"`
}

// SubscriptionCustomerResponse represents the customer data within a subscription response
type SubscriptionCustomerResponse struct {
	ID                 uuid.UUID              `json:"id"`
	NumID              int64                  `json:"num_id"`
	Name               string                 `json:"name,omitempty"`
	Email              string                 `json:"email"`
	Phone              string                 `json:"phone,omitempty"`
	Description        string                 `json:"description,omitempty"`
	FinishedOnboarding bool                   `json:"finished_onboarding"`
	Metadata           map[string]interface{} `json:"metadata,omitempty"`
	CreatedAt          time.Time              `json:"created_at"`
	UpdatedAt          time.Time              `json:"updated_at"`
}

// SubscriptionLineItemResponse represents a line item in a subscription
type SubscriptionLineItemResponse struct {
	ID                   uuid.UUID              `json:"id"`
	SubscriptionID       uuid.UUID              `json:"subscription_id"`
	ProductID            uuid.UUID              `json:"product_id"`
	LineItemType         string                 `json:"line_item_type"` // 'base' or 'addon'
	Quantity             int32                  `json:"quantity"`
	UnitAmountInPennies  int32                  `json:"unit_amount_in_pennies"`
	Currency             string                 `json:"currency"`
	PriceType            string                 `json:"price_type"`
	IntervalType         *string                `json:"interval_type,omitempty"`
	TotalAmountInPennies int32                  `json:"total_amount_in_pennies"`
	IsActive             bool                   `json:"is_active"`
	Metadata             map[string]interface{} `json:"metadata,omitempty"`
	CreatedAt            time.Time              `json:"created_at"`
	UpdatedAt            time.Time              `json:"updated_at"`
	Product              *ProductResponse       `json:"product,omitempty"`
}

// SubscriptionResponse represents a subscription along with its associated product details.
type SubscriptionResponse struct {
	ID                     uuid.UUID                      `json:"id"`
	NumID                  int64                          `json:"num_id"`
	WorkspaceID            uuid.UUID                      `json:"workspace_id"`
	CustomerID             uuid.UUID                      `json:"customer_id,omitempty"`
	CustomerName           string                         `json:"customer_name,omitempty"`
	CustomerEmail          string                         `json:"customer_email,omitempty"`
	Customer               *SubscriptionCustomerResponse  `json:"customer,omitempty"`
	Status                 string                         `json:"status"`
	CurrentPeriodStart     time.Time                      `json:"current_period_start"`
	CurrentPeriodEnd       time.Time                      `json:"current_period_end"`
	NextRedemptionDate     *time.Time                     `json:"next_redemption_date,omitempty"`
	TotalRedemptions       int32                          `json:"total_redemptions"`
	TotalAmountInCents     int32                          `json:"total_amount_in_cents"`
	TokenAmount            int32                          `json:"token_amount"`
	DelegationID           uuid.UUID                      `json:"delegation_id"`
	CustomerWalletID       *uuid.UUID                     `json:"customer_wallet_id,omitempty"`
	ExternalID             string                         `json:"external_id,omitempty"`
	PaymentSyncStatus      string                         `json:"payment_sync_status,omitempty"`
	PaymentSyncedAt        *time.Time                     `json:"payment_synced_at,omitempty"`
	PaymentSyncVersion     int32                          `json:"payment_sync_version,omitempty"`
	PaymentProvider        string                         `json:"payment_provider,omitempty"`
	InitialTransactionHash string                         `json:"initial_transaction_hash,omitempty"`
	Metadata               map[string]interface{}         `json:"metadata,omitempty"`
	CreatedAt              time.Time                      `json:"created_at"`
	UpdatedAt              time.Time                      `json:"updated_at"`
	TrialStart             *time.Time                     `json:"trial_start,omitempty"`
	TrialEnd               *time.Time                     `json:"trial_end,omitempty"`
	CanceledAt             *time.Time                     `json:"canceled_at,omitempty"`
	CancelAt               *time.Time                     `json:"cancel_at,omitempty"`
	CancellationReason     string                         `json:"cancellation_reason,omitempty"`
	PausedAt               *time.Time                     `json:"paused_at,omitempty"`
	PauseEndsAt            *time.Time                     `json:"pause_ends_at,omitempty"`
	ProductID              uuid.UUID                      `json:"product_id"`
	ProductTokenID         uuid.UUID                      `json:"product_token_id"`
	Product                *ProductResponse               `json:"product,omitempty"`
	ProductToken           *ProductTokenResponse          `json:"product_token,omitempty"`
	LineItems              []SubscriptionLineItemResponse `json:"line_items,omitempty"`
}

// GetRedemptionStatusResponse represents the response for the redemption status endpoint
type GetRedemptionStatusResponse struct {
	SubscriptionID   string     `json:"subscription_id"`
	Status           string     `json:"status"` // pending, success, failed
	Message          string     `json:"message"`
	LastRedemptionAt *time.Time `json:"last_redemption_at,omitempty"`
	LastAttemptedAt  *time.Time `json:"last_attempted_at,omitempty"`
	TotalRedemptions int32      `json:"total_redemptions"`
	NextRedemptionAt time.Time  `json:"next_redemption_at"`
	TransactionHash  string     `json:"transaction_hash,omitempty"`
	FailureReason    string     `json:"failure_reason,omitempty"`
}

// SubscribeToProductParams contains all parameters for creating a subscription
type SubscribeToProductParams struct {
	ProductID                 uuid.UUID
	SubscriberAddress         string
	ProductTokenID            string
	TokenAmount               string
	DelegationData            params.DelegationParams
	CypheraSmartWalletAddress string
	Addons                    []SubscriptionAddonParams
}

// SubscriptionAddonParams contains parameters for a subscription addon
type SubscriptionAddonParams struct {
	ProductID uuid.UUID
	Quantity  int32
}

// SubscribeToProductResult contains the result of subscription creation
type SubscribeToProductResult struct {
	Subscription *db.Subscription
	Success      bool
	ErrorMessage string
}

// Deprecated: Use SubscribeToProductParams instead
type SubscribeToProductByPriceIDParams = SubscribeToProductParams

// Deprecated: Use SubscribeToProductResult instead
type SubscribeToProductByPriceIDResult = SubscribeToProductResult
